kube/src/client/mod.rs

```diff
@@ -412,19 +429,35 @@ impl TryFrom<Config> for Client {
         };
 
         let common = ServiceBuilder::new()
-            .map_request(move |r| set_cluster_url(r, &cluster_url))
+            .layer(SetBaseUriLayer::new(cluster_url))
             .map_request(move |r| set_default_headers(r, default_headers.clone()))
             .into_inner();
 
         #[cfg(feature = "gzip")]
         let common = ServiceBuilder::new()
             .layer(common)
-            .map_request(accept_compressed)
-            .map_response(maybe_decompress)
+            .layer(tower_http::decompression::DecompressionLayer::new())
             .into_inner();
 
-        let https: HttpsConnector<_> = config.try_into()?;
-        let mut connector = TimeoutConnector::new(https);
+        let mut connector = HttpConnector::new();
+        connector.enforce_http(false);
+
+        // Note that if both `native_tls` and `rustls` is enabled, `native_tls` is used by default.
+        // To use `rustls`, disable `native_tls` or create custom client.
+        // If tls features are not enabled, http connector will be used.
```

Comment:
While I am happy that we can do without `static_assertions`, I'm a little sad that the build-time; one-tls-stack-dependency guarantee goes away. But it does seem to be the popular demand (even if you end up with bigger executables), and managing build deps can be a complex problem by itself.

However, this does mean that if rustls gets their dns game together (that one caveat listed for rustls in the readme), and we consider seriously switching to rustls as the default, then we now have a new breaking change for it.

I think it's still probably sensible though. It does mean we can do all-features builds again, and generate docs for that.

Reply:
Removing TLS stack requirement is necessary to let users build libraries on top of `kube`. Otherwise, they'll end up having the awkward TLS features like `kube-runtime` did.

I honestly don't know how users using `kube` directly can end up with both enabled _accidentally_ which `static_assertions` prevents. Also, having `native-tls` by default and requiring `default-features = false` is very common. If users are accidentally enabling both often, then we should improve the documentation.

> we consider seriously switching to rustls as the default

I do want rustls to become the default eventually, but I don't think this will happen any time soon. Validating certificates for IP addresses is not the only issue. Support for different types of private keys are lacking. For example, EC Private Key used by k3d is not supported, so rustls fails to find the key (only PKCS#8 and RSA (PKCS#1) can be extracted from PEM). To use rustls with k3d, we need to change the cluster's host to `localhost` and convert the private key in kubeconfig to PKCS#8 using OpenSSL. We can do this in `kube`, but OpenSSL will be required :(

Native TLS connector only supports loading Identity from PKCS#12, but we're using OpenSSL to create the archive, so the conversions is done there. This is also not ideal because of OpenSSL dependency on macOS and Windows, especially the latter is painful. It's possible to create PKCS#12 archive in pure Rust, but some key formats won't be supported.

Reply:
Yeah, I definitely think this is the right thing to do.

For some of these rustls caveats, it might be worth creating another tracking issue so that this is not lost. I thought it was only the IP issue :-(

Reply:
Opened #542